File: Engine/ai.py
import random

def _get_potential_cells(game, color):
    """
    Возвращает множество клеток, которые могут участвовать в квадратах с фишками `color`.
    Это:
      - Все пустые клетки, которые вместе с 3 фишками color образуют квадрат (выигрыш)
      - Все пустые клетки, которые вместе с 2 фишками color могут стать квадратом
    """
    n = game.n
    board = game.board
    points = [(i, j) for i in range(n) for j in range(n) if board[i][j] == color]
    potential = set()

    if len(points) < 2:
        return set()

    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            A = points[i]
            B = points[j]
            ax, ay = A
            bx, by = B

            dx = bx - ax
            dy = by - ay

            C1 = (ax - dy, ay + dx)
            D1 = (bx - dy, by + dx)

            C2 = (ax + dy, ay - dx)
            D2 = (bx + dy, by - dx)

            candidates = [C1, D1, C2, D2]
            for pt in candidates:
                x, y = pt
                if 0 <= x < n and 0 <= y < n and board[x][y] == '.':
                    potential.add((x, y))

    return potential


def get_best_move(game):
    """
    Возвращает лучший ход, используя только потенциально значимые клетки.
    Логика:
      1. Выигрышный ход
      2. Блокировка
      3. Среди потенциальных клеток — выбрать ту, что максимизирует наши квадраты / минимизирует чужие
    """
    current_color = game.get_player_color()
    opponent_color = 'B' if current_color == 'W' else 'W'
    n = game.n

    all_empty = [(i, j) for i in range(n) for j in range(n) if game.board[i][j] == '.']
    if not all_empty:
        return None

    for x, y in all_empty:
        game.board[x][y] = current_color
        if game.check_win(x, y, current_color):
            game.board[x][y] = '.'
            return (x, y)
        game.board[x][y] = '.'

    for x, y in all_empty:
        game.board[x][y] = opponent_color
        if game.check_win(x, y, opponent_color):
            game.board[x][y] = '.'
            return (x, y)
        game.board[x][y] = '.'

    my_potential = _get_potential_cells(game, current_color)
    opp_potential = _get_potential_cells(game, opponent_color)
    
    candidate_cells = my_potential | opp_potential

    if candidate_cells:
        best_score = -10**9
        best_move = None

        for x, y in candidate_cells:
            score = 0

            game.board[x][y] = current_color

            if (x, y) in my_potential:
                score += 2
            if (x, y) in opp_potential:
                score -= 1

            game.board[x][y] = '.'

            if score > best_score:
                best_score = score
                best_move = (x, y)

        if best_move:
            return best_move

    return random.choice(all_empty)

File: Engine/src/main/java/ai/AI.java
package ai;

import game.Board;
import game.Game;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

public class AI {
    public static int[] getBestMove(Game game) {
        Board board = game.getBoard();
        char currentColor = game.getCurrentPlayerColor();
        char opponentColor = (char)(currentColor == 'W' ? 66 : 87);
        List<int[]> emptyCells = board.getEmptyCells();
        if (emptyCells.isEmpty()) {
            return null;
        } else {
            for(int[] cell : emptyCells) {
                if (board.wouldWin(cell[0], cell[1], currentColor)) {
                    return cell;
                }
            }

            for(int[] cell : emptyCells) {
                if (board.wouldWin(cell[0], cell[1], opponentColor)) {
                    return cell;
                }
            }

            Set<int[]> myPotential = getPotentialCells(board, currentColor);
            Set<int[]> oppPotential = getPotentialCells(board, opponentColor);
            int bestScore = Integer.MIN_VALUE;
            int[] bestMove = null;

            for(int[] cell : emptyCells) {
                int score = 0;
                if (containsPoint(myPotential, cell)) {
                    score += 2;
                }

                if (containsPoint(oppPotential, cell)) {
                    --score;
                }

                if (score > bestScore) {
                    bestScore = score;
                    bestMove = cell;
                }
            }

            return bestMove != null ? bestMove : randomChoice(emptyCells);
        }
    }

    private static boolean containsPoint(Set<int[]> set, int[] point) {
        for(int[] p : set) {
            if (Arrays.equals(p, point)) {
                return true;
            }
        }

        return false;
    }

    private static int[] randomChoice(List<int[]> list) {
        return (int[])list.get((new Random()).nextInt(list.size()));
    }

    private static Set<int[]> getPotentialCells(Board board, char color) {
        int n = board.getSize();
        List<int[]> points = new ArrayList();

        for(int i = 0; i < n; ++i) {
            for(int j = 0; j < n; ++j) {
                if (board.getCell(i, j) == color) {
                    points.add(new int[]{i, j});
                }
            }
        }

        Set<int[]> potential = new HashSet();
        if (points.size() < 2) {
            return potential;
        } else {
            for(int i = 0; i < points.size(); ++i) {
                for(int j = i + 1; j < points.size(); ++j) {
                    int[] A = (int[])points.get(i);
                    int[] B = (int[])points.get(j);
                    int ax = A[0];
                    int ay = A[1];
                    int bx = B[0];
                    int by = B[1];
                    int dx = bx - ax;
                    int dy = by - ay;
                    int[][] candidates = new int[][]{{ax - dy, ay + dx}, {bx - dy, by + dx}, {ax + dy, ay - dx}, {bx + dy, by - dx}};

                    for(int[] pt : candidates) {
                        int x = pt[0];
                        int y = pt[1];
                        if (x >= 0 && x < n && y >= 0 && y < n && board.getCell(x, y) == '.') {
                            potential.add(new int[]{x, y});
                        }
                    }
                }
            }

            return potential;
        }
    }
}
